README.txt
= sogoumap2pdf

* sogoumap2pdf.rubyforge.org

== DESCRIPTION:

* Download map and convert to pdf

== FEATURES/PROBLEMS:

* Download map and convert to pdf


== SYNOPSIS:
  Find out the link of the center picture using Firebug plugin in Firefox, then type:
    sogoumap2pdf link map_widthxmap_height rowxcolumn save_file_name mode
  for example:
    sogoumap2pdf http://hbpic1.go2map.com/seamless/0/180/711/252/52/50449_10516.JPG 1024x768 5x5 gz.pdf sp
  mode description:
    sp: both satellite and path
    s: only satellite
    m: only map

  Links for some cities:
    1 Beijing:http://hbpic3.go2map.com/seamless/0/180/792/518/192/103647_38598.JPG
    2 Shanghai:http://hbpic3.go2map.com/seamless/0/180/792/540/145/108175_29128.JPG
    3 Tianjing:http://hbpic3.go2map.com/seamless/0/180/792/521/188/104367_37713.JPG
    4 Chongqin:http://hbpic2.go2map.com/seamless/0/180/792/474/137/94897_27407.JPG
    5 Chendu:http://hbpic4.go2map.com/seamless/0/180/792/463/142/92685_28539.JPG
    6 Dalian:http://hbpic2.go2map.com/seamless/0/180/792/541/187/108301_37462.JPG
    7 Fuzhou:http://hbpic0.go2map.com/seamless/0/180/792/531/119/106244_23925.JPG
    8 Guangzhou:http://hbpic0.go2map.com/seamless/0/180/792/504/105/100915_21050.JPG
    9 Hangzhou:http://hbpic1.go2map.com/seamless/0/180/792/535/140/107001_28156.JPG
    10 Jinan:http://hbpic4.go2map.com/seamless/0/180/792/520/174/104176_34934.JPG
    11 Kunming:http://hbpic0.go2map.com/seamless/0/180/792/457/114/91472_22905.JPG
    12 Nanjing:http://hbpic1.go2map.com/seamless/0/180/792/528/149/105778_29981.JPG
    13 Qingdao:http://hbpic0.go2map.com/seamless/0/180/792/536/171/107220_34345.JPG
    14 Sanya:http://hbpic4.go2map.com/seamless/0/180/792/487/82/97523_16429.JPG
    15 Shenzhen:http://hbpic4.go2map.com/seamless/0/180/792/507/102/101555_20499.JPG
    16 Shenyan:http://hbpic0.go2map.com/seamless/0/180/792/549/204/109907_40820.JPG
    17 Wuhan:http://hbpic3.go2map.com/seamless/0/180/792/508/142/101779_28448.JPG
    18 Xian:http://hbpic4.go2map.com/seamless/0/180/792/485/161/97019_32314.JPG
    19 Hongkong:http://hbpic3.go2map.com/seamless/0/180/792/508/101/101692_20253.JPG
    20 Zhengzhou:http://hbpic4.go2map.com/seamless/0/180/792/506/164/101221_32844.JPG
    21 Qinghuangdao:http://hbpic3.go2map.com/seamless/0/180/792/532/193/106514_38638.JPG
== REQUIREMENTS:

* pdf-writer
* rmagick(On ubuntu, imagemagick and libmagick9-dev are needed to be installed first)

== INSTALL:

* sudo gem install sogoumap2pdf

== LICENSE:

(The MIT License)

Copyright (c) 2008 FIXME full name

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 0.0.1 2008-08-19

* 1 major enhancement:
  * Initial release
== 0.0.2 2008-08-19

* 1 major enhancement:
  * Import example

== 0.0.2 2008-08-19
* 1 major enhancement:
  * Change command syntax

== 0.0.8 2008-08-19
* 1 bug fixed:
  * Page size error

== 0.0.9 2008-08-19
* 1 minor enhancement:
  * add command sogoumap2disk

== 0.0.11 2008-09-09
* 1 minor enhancement:
  * add command sogoumap2pdf_from_disk

== 0.1.0 2008-09-10
* 1 bug fixed:
  fix crop! error
== 0.2.0 2008-09-11
* 1 bug fixed:
  fix front page error


spec/spec_helper.rb
begin
  require 'spec'
rescue LoadError
  require 'rubygems'
  gem 'rspec'
  require 'spec'
end

$:.unshift(File.dirname(__FILE__) + '/../lib')
require 'sogoumap2pdf'


config/hoe.rb
require 'sogoumap2pdf/version'

AUTHOR = 'David Ruan'  # can also be an array of Authors
EMAIL = "ruanwz@gmail.com"
DESCRIPTION = "download the map and convert to pdf"
GEM_NAME = 'sogoumap2pdf' # what ppl will type to install your gem
RUBYFORGE_PROJECT = 'sogoumap2pdf' # The unix name for your project
HOMEPATH = "http://#{RUBYFORGE_PROJECT}.rubyforge.org"
DOWNLOAD_PATH = "http://rubyforge.org/projects/#{RUBYFORGE_PROJECT}"
EXTRA_DEPENDENCIES = [
   ['pdf-writer', '>= 1.1.8'],
   ['rmagick', '>= 2.5.2']
]    # An array of rubygem dependencies [name, version]

@config_file = "~/.rubyforge/user-config.yml"
@config = nil
RUBYFORGE_USERNAME = "ruanwz"
def rubyforge_username
  unless @config
    begin
      @config = YAML.load(File.read(File.expand_path(@config_file)))
    rescue
      puts <<-EOS
ERROR: No rubyforge config file found: #{@config_file}
Run 'rubyforge setup' to prepare your env for access to Rubyforge
 - See http://newgem.rubyforge.org/rubyforge.html for more details
      EOS
      exit
    end
  end
  RUBYFORGE_USERNAME.replace @config["username"]
end


REV = nil
# UNCOMMENT IF REQUIRED:
# REV = YAML.load(`svn info`)['Revision']
VERS = Sogoumap2pdf::VERSION::STRING + (REV ? ".#{REV}" : "")
RDOC_OPTS = ['--quiet', '--title', 'sogoumap2pdf documentation',
    "--opname", "index.html",
    "--line-numbers",
    "--main", "README",
    "--inline-source"]

class Hoe
  def extra_deps
    @extra_deps.reject! { |x| Array(x).first == 'hoe' }
    @extra_deps
  end
end

# Generate all the Rake tasks
# Run 'rake -T' to see list of generated tasks (from gem root directory)
$hoe = Hoe.new(GEM_NAME, VERS) do |p|
  p.developer(AUTHOR, EMAIL)
  p.description = DESCRIPTION
  p.summary = DESCRIPTION
  p.url = HOMEPATH
  p.rubyforge_name = RUBYFORGE_PROJECT if RUBYFORGE_PROJECT
  p.test_globs = ["test/**/test_*.rb"]
  p.clean_globs |= ['**/.*.sw?', '*.gem', '.config', '**/.DS_Store']  #An array of file patterns to delete on clean.

  # == Optional
  p.changes = p.paragraphs_of("History.txt", 0..1).join("\n\n")
  p.extra_deps = EXTRA_DEPENDENCIES

    #p.spec_extras = {}    # A hash of extra values to set in the gemspec.
  end

CHANGES = $hoe.paragraphs_of('History.txt', 0..1).join("\\n\\n")
PATH    = (RUBYFORGE_PROJECT == GEM_NAME) ? RUBYFORGE_PROJECT : "#{RUBYFORGE_PROJECT}/#{GEM_NAME}"
$hoe.remote_rdoc_dir = File.join(PATH.gsub(/^#{RUBYFORGE_PROJECT}\/?/,''), 'rdoc')
$hoe.rsync_args = '-av --delete --ignore-errors'
$hoe.spec.post_install_message = File.open(File.dirname(__FILE__) + "/../PostInstall.txt").read rescue ""


website/ubiquity-torrent.js
var MCM = {
'ANIME' : '1',
'BOOKS' : '2',
'GAMES' : '3',
'MOVIES' : '4',
'MUSIC' : '5',
'PICTURES' : '6',
'SOFTWARE' : '7',
'TV SHOWS' : '8',
'OTHER' : '9'
};

var MininovaCategories = [
'Anime',
'Books',
'Games',
'Movies',
'Music',
'Pictures',
'Software',
'TV shows',
'Other'
];

var noun_type_mincats= new CmdUtils.NounType( "Category", MininovaCategories );

CmdUtils.CreateCommand({
  name: "mininova",
  author: { name: "Gary Hodgson", homepage : "http://www.garyhodgson.com/", email : "contact@garyhodgson.com"},
  icon: "http://static.mininova.org/images/favicon.ico",
  homepage: "http://garyhodgson/ubiquity",
  license: "MPL",
  releaseinfo: {2:"(01 Sep 2008) Replaced openUrl with Utils.openUrlInBrowser.",
  				1:"(29 Aug 2008) Initial Release." },
  description: "Performs a search on mininova.",
  help: "A category can be defined to limit the result set, e.g. 'in Books'",

  takes: {"torrent" : noun_arb_text},
  modifiers: {"in" : noun_type_mincats},
  
  preview: function( pblock, searchText, mods ) {
    pblock.innerHTML = "";
    var msg = 'Search Mininova for : ${search} in ${category}, and sort by most seeders.';
    var subs = {search: searchText.text, category: (mods.in.text||'')};
    
   pblock.innerHTML = CmdUtils.renderTemplate( msg, subs ); 
 
  },
  
execute: function( searchText, mods ) {
  
  h='www.mininova.org';
  var category = mods.in.text||'';
  var catid = ( MCM[category.toUpperCase()] ? '/'+MCM[category.toUpperCase()] : '');
  p='/search/'+encodeURIComponent(searchText.text) + catid + '/seeds';
  
  cp='http://'+h+p;
  Utils.openUrlInBrowser(cp,null);
}
  
});

CmdUtils.CreateCommand({
  name: "mininova-imdb",
  author: { name: "Gary Hodgson", homepage : "http://www.garyhodgson.com/", email : "contact@garyhodgson.com"},
  icon: "http://static.mininova.org/images/favicon.ico",
  homepage: "http://garyhodgson/ubiquity",
  license: "MPL",
  releaseinfo: {2:"(01 Sep 2008) Replaced openUrl with Utils.openUrlInBrowser.",
  			 	1:"(29 Aug 2008) Initial Release."},
  description: "Searches mininova for torrents associated with the given IMDB reference.",
  help: "IMDB references can include the tt prefix or not.",
  takes: {"imdbid" : noun_arb_text},
  
  preview: function( pblock, searchText, mods ) {
    pblock.innerHTML = "";
    var msg = 'Search Mininova for : ${imdbid} , using an IMDB reference.';
    var subs = {imdbid: searchText.text};
    
   pblock.innerHTML = CmdUtils.renderTemplate( msg, subs ); 
 
  },
  
execute: function( searchText, mods ) {
  
  h='www.mininova.org';
  var imdbid = ( (! searchText.text.indexOf("tt")) ? searchText.text.slice(2) : searchText.text);
  p='/imdb/?imdb='+imdbid;
  cp='http://'+h+p;
  Utils.openUrlInBrowser(cp,null);
}
  
});

CmdUtils.CreateCommand({
  name: "myebook",
  author: { name: "David Ruan", homepage : "http://sogoumap2pdf.rubyforge.org/", email : "ruanwz@gmail.com"},
  icon: "http://static.mininova.org/images/favicon.ico",
  homepage: "http://sogoumap2pdf.rubyforge.org/",
  license: "MPL",
  releaseinfo: {2:"(04 Sep 2008) Fix bugs.",
  			 	1:"(29 Aug 2008) Initial Release."},
  description: "Searches ebook on ITPUB,then finds comments on douban.",
  help: "Pass the book name.",
  takes: {"book_name" : noun_arb_text},
  
  preview: function( pblock, searchText, mods ) {
    pblock.innerHTML = "";
    var msg = 'Search ITPUB for : ${book_name} , using a book name.';
    var subs = {book_name: searchText.text};
    debugger 
    pblock.innerHTML = CmdUtils.renderTemplate( msg, subs ); 
 
  },
  
execute: function( searchText, mods ) {
  var head='www.itpub.net/search.php';
  var book_name = searchText.text;
  var parms = {srchfid: 61, srchtext: book_name};
  debugger 
  jQuery.post(head,parms,function(resp){
  CmdUtils.setSelection(resp.text);
  })
}
  
});


lib/sogoumap2pdf/image_to_pdf.rb
require 'rubygems'
require 'pdf/writer'
require 'net/http'
require 'uri'
require 'RMagick'
include Magick

module Sogoumap2pdf
  class ImageToPdf
    def initialize(link_end,page_width,page_height)
      #@page_width=600
      @page_width=page_width.to_i
      #@page_width=800
      #@page_height=768
      @page_height=page_height.to_i
      #@page_height=480
      @page_overlap=50
      @text_height=15
      @pdf=PDF::Writer.new(:paper=>[0,0,@page_width,@page_height])
      @pdf.margins_pt(0,0,0,0)
      # initialize icons
      @icon_size=20
      @link_end=link_end
    end
    def big_img_crop(x,y,w,h)
      g=@g
      nx0=x/256
      ny0=y/256
      nxn=(x+w)/256+1
      nyn=(y+h)/256+1
      delta_x=x-nx0*256
      delta_y=y-ny0*256
      # p "x #{x}"
      # p "y #{y}"
      # p "delta_x #{delta_x}"
      # p "delta_y #{delta_y}"
      ci=ImageList.new # image list to hold results
      page=Rectangle.new( 0, 0, 0, 0)
  
      for i in nx0..nxn
        for j in ny0..nyn
          #p "big_img_crop #{i},#{j}"
          picture_file_name="~/.sogoumap/0/"+ [g.resource_type,g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+g.link_end
          picture_file_name=File.expand_path(picture_file_name)
          if File.exist?(picture_file_name) && File.size(picture_file_name)!= 0 then
            tumb=ImageList.new.read(picture_file_name)[0] # image list to hold results
            path_file_name="~/.sogoumap/0/"+ ["179",g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+".PNG"
            path_file_name=File.expand_path(path_file_name)
  
            if g.satellite_path && File.exist?(path_file_name) && File.size(path_file_name)!=0 then
  
              path_tumb=ImageList.new.read(path_file_name)[0] # image list to hold results
              tumb.composite!(path_tumb, 0, 0, OverCompositeOp)
            end
  
            #append tumb
            ci << tumb
            #update page
            page.x=(i-nx0)*256
            page.y=(nyn-j)*256
            ci.page=page
          else
          end
        end
      end
      if ci.length==0 then
        resimg=ImageList.new.read("NULL:white") { self.size = "#{g.npx*@tumb_x}x#{g.npy*@tumb_y}"}
      else
        resimg=ci.mosaic
      end
      return resimg.crop(SouthWestGravity,delta_x,delta_y,w,h,true)
    end
  
    def perfect_fit(w,h,r,c)
      @page_width=w
      @page_height=h
      @row=r
      @column=c
      @tumb_x=@page_width*256/(c*@page_width-(c-1)*@page_overlap)
      @tumb_y=@page_height*256/(r*@page_height-(r-1)*@page_overlap)
      @big_img_width,@big_img_height=c*@page_width-(c-1)*@page_overlap, r*@page_height-(r-1)*@page_overlap
      return c*@page_width-(c-1)*@page_overlap, r*@page_height-(r-1)*@page_overlap
    end
    def create_pages(g)
      # get the size of the image
      size_x=@big_img_width
      size_y=@big_img_height
  
      # calculate the number of requred rows and columns
      cols=(size_x-@page_overlap)/(@page_width-@page_overlap)
      rows=(size_y-@page_overlap)/(@page_height-@page_overlap)
      # create pdf
      for i in 0...rows
        for j in 0...cols
          p "creating page #{i} #{j}"
          p Time.now
          @pdf.new_page
          @pdf.add_destination("map part #{i},#{j}", "Fit")
          pgimg=big_img_crop(j*(@page_width-@page_overlap),
                               (rows-1-i)*(@page_height-@page_overlap),@page_width,@page_height)
  
          # pgimg=big_image.crop(j*(@page_width-@page_overlap),
          #                      i*(@page_height-@page_overlap),@page_width,@page_height,true)
          pgimg.write("itptmp#{@link_end}") { self.quality = 100 }
          @pdf.add_image_from_file("itptmp#{@link_end}",0,0)
          ss = PDF::Writer::StrokeStyle.new(4,:cap=>:round)
          @pdf.stroke_style ss
          if j>0 then
            @pdf.add_internal_link("map part #{i},#{j-1}",0,@page_height/2,
                                   @icon_size,@page_height/2+@icon_size)
            @pdf.line(@icon_size,@page_height/2,0,@page_height/2+@icon_size/2).stroke
            @pdf.line(0,@page_height/2+@icon_size/2,@icon_size,@page_height/2+@icon_size).stroke
          end
          if j<cols-1 then
            @pdf.add_internal_link("map part #{i},#{j+1}",
                                   @page_width-@icon_size,@page_height/2,
                                   @page_width,@page_height/2+@icon_size)
            @pdf.line(@page_width-@icon_size,@page_height/2,
                      @page_width,@page_height/2+@icon_size/2).stroke
            @pdf.line(@page_width,@page_height/2+@icon_size/2,
                      @page_width-@icon_size,@page_height/2+@icon_size).stroke
          end
          if i>0 then
            @pdf.add_internal_link("map part #{i-1},#{j}",@page_width/2,@page_height,
                                   @page_width/2+@icon_size,@page_height-@icon_size)
            @pdf.line(@page_width/2,@page_height-@icon_size,
                      @page_width/2+@icon_size/2,@page_height).stroke
            @pdf.line(@page_width/2+@icon_size/2,@page_height,
                      @page_width/2+@icon_size,@page_height-@icon_size).stroke
          end
          if i<rows-1 then
            @pdf.add_internal_link("map part #{i+1},#{j}",@page_width/2,0,
                                    @page_width/2+@icon_size,@icon_size)
            @pdf.line(@page_width/2,@icon_size,
                      @page_width/2+@icon_size/2,0).stroke
            @pdf.line(@page_width/2+@icon_size/2,0,
                      @page_width/2+@icon_size,@icon_size).stroke
          end
  #        @pdf.restore_state          
        end
      end
    end
    def front_page(g)
      size_x=@big_img_width
      size_y=@big_img_height
      # calculate the number of requred rows and columns
      cols=(size_x-@page_overlap)/(@page_width-@page_overlap)
      rows=(size_y-@page_overlap)/(@page_height-@page_overlap)
  
      ci=ImageList.new # image list to hold results
      page=Rectangle.new( 0, 0, 0, 0)
      @g=g
      for i in 0... g.npx
        p "resizing #{i} j"
        for j in 0...g.npy
          picture_file_name="~/.sogoumap/0/"+ [g.resource_type,g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+g.link_end
          picture_file_name=File.expand_path(picture_file_name)
          if File.exist?(picture_file_name) && File.size(picture_file_name)!= 0 then
            tumb=ImageList.new.read(picture_file_name)[0] # image list to hold results
  
            path_file_name="~/.sogoumap/0/"+ ["179",g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+".PNG"
            path_file_name=File.expand_path(path_file_name)
            if g.satellite_path && File.exist?(path_file_name) && File.size(path_file_name)!=0 then
              path_tumb=ImageList.new.read(path_file_name)[0] # image list to hold results
              tumb.composite!(path_tumb, 0, 0, OverCompositeOp)
            end
            #resize each to tumb
            tumb.resize!(@tumb_x,@tumb_y)
            #append tumb
            ci << tumb
            #update page
            page.x=i*@tumb_x
            page.y=(g.npy-1-j)*@tumb_y
            ci.page=page
          else
          end
        end
      end
      #mosaic
      if ci.length==0 then
        p "@tumb_x,@tumb_y"
        p @tumb_x
        p @tumb_y
        p "g.npx,g.npy"
        p g.npx
        p g.npy
        p "#{g.npx*@tumb_x}x#{g.npy*@tumb_y}"
        ## can't work
        resimg=ImageList.new.read("null:white") { self.size = "#{g.npx.to_i*@tumb_x.to_i}x#{g.npy.to_i*@tumb_y.to_i}"}
      else
        resimg=ci.mosaic
      end
      #crop
      #map2pdf.rb:339:in `crop!': bignum too big to convert into `long' (RangeError)
      resx=@tumb_x*g.npx
      resy=@tumb_y*g.npy

      p "front_page size"
      p resx
      p resy
      resx=@page_width if resx>@page_width
      resy=@page_height if resy>@page_height
      resimg.crop!(SouthWestGravity,resx,resy)
      resimg.write("itptmp#{@link_end}"){ self.quality = 100 }
      @pdf.add_image_from_file("itptmp#{@link_end}",0,0)
      ss = PDF::Writer::StrokeStyle.new(2)
      @pdf.stroke_style ss
      for i in 1...cols
        @pdf.line(i*@page_width/cols,0,
                  i*@page_width/cols,@page_height).stroke
      end
      for i in 1...rows
        @pdf.line(0,i*@page_height/rows,
                  @page_width,i*@page_height/rows).stroke
      end
      for i in 0...rows
        for j in 0...cols
         p "adding #{i} #{j} link in front page"
         p Time.now
         @pdf.add_internal_link("map part #{i},#{j}",
                                j*@page_width/cols, @page_height-i*@page_height/rows,
                                (j+1)*@page_width/cols, @page_height-(i+1)*@page_height/rows)
       end
      end
    end
    def front_page_only(g)
      size_x=@big_img_width
      size_y=@big_img_height
      # calculate the number of requred rows and columns
      cols=(size_x-@page_overlap)/(@page_width-@page_overlap)
      rows=(size_y-@page_overlap)/(@page_height-@page_overlap)
  
      ci=ImageList.new # image list to hold results
      page=Rectangle.new( 0, 0, 0, 0)
      @g=g
      for i in 0... g.npx
        p "resizing #{i} j"
        for j in 0...g.npy
          picture_file_name="~/.sogoumap/0/"+ [g.resource_type,g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+g.link_end
          picture_file_name=File.expand_path(picture_file_name)
          if File.exist?(picture_file_name) && File.size(picture_file_name)!= 0 then
            tumb=ImageList.new.read(picture_file_name)[0] # image list to hold results
  
            path_file_name="~/.sogoumap/0/"+ ["179",g.resource_level,((g.start_link_x+i)/200).to_s,((g.start_link_y+j)/200).to_s].join("/")+"/"+(g.start_link_x+i).to_s+"_"+(g.start_link_y+j).to_s+".PNG"
            path_file_name=File.expand_path(path_file_name)
            if g.satellite_path && File.exist?(path_file_name) && File.size(path_file_name)!=0 then
              path_tumb=ImageList.new.read(path_file_name)[0] # image list to hold results
              tumb.composite!(path_tumb, 0, 0, OverCompositeOp)
            end
            #resize each to tumb
            tumb.resize!(@tumb_x,@tumb_y)
            #append tumb
            ci << tumb
            #update page
            page.x=i*@tumb_x
            page.y=(g.npy-1-j)*@tumb_y
            ci.page=page
          else
          end
        end
      end
      #mosaic
      if ci.length==0 then
        p "@tumb_x,@tumb_y"
        p @tumb_x
        p @tumb_y
        p "g.npx,g.npy"
        p g.npx
        p g.npy
        p "#{g.npx*@tumb_x}x#{g.npy*@tumb_y}"
        ## can't work
        resimg=ImageList.new.read("null:white") { self.size = "#{g.npx.to_i*@tumb_x.to_i}x#{g.npy.to_i*@tumb_y.to_i}"}
      else
        resimg=ci.mosaic
      end
      #crop
      #map2pdf.rb:339:in `crop!': bignum too big to convert into `long' (RangeError)
      resx=@tumb_x*g.npx
      resy=@tumb_y*g.npy
      p "front_page size"
      p resx
      p resy
      resx=@page_width if resx>@page_width
      resy=@page_height if resy>@page_height
      resimg.crop!(SouthWestGravity,resx,resy)
      resimg.write("itptmp#{@link_end}"){ self.quality = 100 }
      @pdf.add_image_from_file("itptmp#{@link_end}",0,0)
      ss = PDF::Writer::StrokeStyle.new(2)
      @pdf.stroke_style ss
      for i in 1...cols
        @pdf.line(i*@page_width/cols,0,
                  i*@page_width/cols,@page_height).stroke
      end
      for i in 1...rows
        @pdf.line(0,i*@page_height/rows,
                  @page_width,i*@page_height/rows).stroke
      end
      # for i in 0...rows
      #   for j in 0...cols
      #    p "adding #{i} #{j} link in front page"
      #    p Time.now
      #    @pdf.add_internal_link("map part #{i},#{j}",
      #                           j*@page_width/cols, @page_height-i*@page_height/rows,
      #                           (j+1)*@page_width/cols, @page_height-(i+1)*@page_height/rows)
      #   end
      # end
    end

    def save(save_file)
      @pdf.save_as(save_file)
    end
  end
end


lib/sogoumap2pdf/sogou_map_image.rb
#!/usr/bin/ruby
require 'rubygems'
require 'pdf/writer'
require 'net/http'
require 'uri'
require 'RMagick'
include Magick

# provides an access to sogou map 
# emulates it as an image
#
# must be initialised by a link to an image from map.sogou.com
# 
# command syntax:
# guangzhou: sogoumap2pdf http://hbpic3.go2map.com/seamless/0/180/792/504/105/100899_21033.JPG 1024x768 3x3 guangzhou.pdf sp
module Sogoumap2pdf
  class SogouMapImage
    attr_reader   :link_end, :satellite_path, :start_link_x, :start_link_y, :picture_directory_name, :path_directory_name, :npx, :npy, :area_width, :area_height,:resource_type,:resource_level,:resource_extension
    def initialize(sogou_link,option)
      @link_beginning, @link_end = sogou_link.split(/\d+_\d+/)
      @sogou_link=sogou_link
      resource_head,resource_parameters= sogou_link.split(/\/0\//)
      @resource_head = resource_head+"/0/"
      @resource_type,@resource_level,@resource_x200,@resource_y200,@resource_x,@resource_y,@resource_extension=resource_parameters.split(/\/|_|\./)
      @link_x=@resource_x.to_i
      @link_y=@resource_y.to_i
  
      @satellite_only=true if option=="s"
      @satellite_path=true if option=="sp"
      @map=true if option=="m"
      @patch_size_x=256
      @patch_size_y=256
      #for sogou map
      #@link_beginning="http://hbpic2.go2map.com/seamless/0/180/792/504/105/"
      #@link_end=".JPG"
      #@link_x=100960
      #@link_y=21035
    end
    # get a patch of map from google maps
    # the initial link is the patch 0,0
    def get_or_read_picture(link_name,dir_name,file_name)
      file_name=File.expand_path(file_name)
      if File.exist? file_name then
        #if File.size(file_name) !=0 then
        #  picture_block=File.read file_name
        #else
        #  picture_block = File.read("empty.jpg") if link_name =~ /(JPG|GIF)$/
        #  picture_block = File.read("empty.png") if link_name =~ /PNG$/
        #end
      else
        picture_data=Net::HTTP.get_response(URI.parse(link_name))
        p picture_data.code
        if picture_data.code == "200" then
          picture_block = picture_data.body
          unless File.exist? dir_name then
            #Dir.mkdir dir_name
            `mkdir -p #{dir_name}`
          end
          tmp_file=File.new(file_name,"w")
                tmp_file.binmode
          tmp_file.write(picture_data.body)
          tmp_file.close
        else
          unless File.exist? dir_name then
            #Dir.mkdir dir_name
            `mkdir -p #{dir_name}`
          end
          `touch #{file_name}`
          #temp_blob = ImageList.new.read("NULL:white") { self.size = "256x256"}
          #picture_block = temp_blob.to_blob
          #picture_block = File.read("empty.jpg") if link_name =~ /(JPG|GIF)$/
          #picture_block = File.read("empty.png") if link_name =~ /PNG$/
        end
      end
      return picture_block
    end
    def only_save_image(x,y)
      http_get_picture = false
      http_get_path = false
      @resource_x200=((@link_x+x)/200).to_s
      @resource_y200=((@link_y+y)/200).to_s
      picture_link=@resource_head+[@resource_type,@resource_level,@resource_x200,@resource_y200].join("/")+"/"+(@link_x+x).to_s+"_"+(@link_y+y).to_s+"."+@resource_extension
      p "getting #{x-@spx}/#{@npx-1}, #{y-@spy}/#{@npy-1} #{picture_link}"
      picture_directory_name = "~/.sogoumap/0/"+ [@resource_type,@resource_level,@resource_x200,@resource_y200].join("/") 
      picture_file_name = picture_directory_name +"/"+ (@link_x+x).to_s+"_"+(@link_y+y).to_s+"."+@resource_extension 
      get_or_read_picture(picture_link,picture_directory_name,picture_file_name)
      if @satellite_path then
        path_link=@resource_head+["179",@resource_level,@resource_x200,@resource_y200].join("/")+"/"+(@link_x+x).to_s+"_"+(@link_y+y).to_s+"."+"PNG"
        p "getting #{x-@spx}/#{@npx-1}, #{y-@spy}/#{@npy-1} #{path_link}"
        path_data=Net::HTTP.get_response(URI.parse(path_link))
        path_directory_name = "~/.sogoumap/0/"+ ["179",@resource_level,@resource_x200,@resource_y200].join("/") 
        path_file_name = path_directory_name + "/" +(@link_x+x).to_s+"_"+(@link_y+y).to_s+".PNG" 
        get_or_read_picture(path_link,path_directory_name, path_file_name)
      end
    end
    # creates map of given size (in pixels), returns Image object
    def fill_map(rx,ry)
      @npx=npx=rx/@patch_size_x+1
      @npy=npy=ry/@patch_size_y+1
      @spacex=spacex=npx*@patch_size_x-rx
      @spacey=spacey=npy*@patch_size_y-ry
      # starting patch
      @spx=spx=-(npx/2)
      @spy=spy=-(npy/2)
      @start_link_x=spx+@link_x
      @start_link_y=spy+@link_y
      @area_height=npy*@patch_size_y
      @area_width=npx*@patch_size_x
      # end patch
      # epx=npx+spx-1
      # epy=npy+spy-1
      # for i in 0...npx
      #   for j in 0...npy
      #     self.only_save_image(i+spx,j+spy)
      #   end
      # end
    end
    def fill_map2
      for i in 0...@npx
        for j in 0...@npy
          self.only_save_image(i+@spx,j+@spy)
        end
      end
    end

  end
end
